### PR TITLE
feat(expr): support VARIADIC function arguments

### LINKED ISSUE #14705: Support VARIADIC function arguments
### Is your feature request related to a problem? Please describe.

While looking for a workaround for https://github.com/risingwavelabs/risingwave/issues/14704 to parse ISO 8601 intervals, I thought to try regular expressions to parse an ISO 8601 string myself.  This seemed easy enough, especially since my source (Debezium) always outputs the intervals in a very rigid format.  A regexp that captures each ISO 8601 interval element, and then uses it in a format string:

**PostgreSQL SQL**

```
users=# select format('%s year %s month %s day %s:%s:%s', variadic regexp_match('P1Y2M3DT4H5M6.15S', 'P(\d+)Y(\d+)M(\d+)DT(\d+)H(\d+)M([\d\.]+)S'));
            format
-------------------------------
 1 year 2 month 3 day 4:5:6.15
```

**RisingWave SQL**

```
thumbtack=> select format('%s year %s month %s day %s:%s:%s', variadic regexp_match('P1Y2M3DT4H5M6.15S', 'P(\d+)Y(\d+)M(\d+)DT(\d+)H(\d+)M([\d\.]+)S'));
ERROR:  Failed to run the query

Caused by:
  sql parser error: syntax error at or near variadic at line:1, column:57
```

It seems that RisingWave does not understand the VARIADIC keyword used by PostgreSQL for unpacking an array into function arguments, as described at https://www.postgresql.org/docs/16/xfunc-sql.html#XFUNC-SQL-VARIADIC-FUNCTIONS

It seems there was a special case recently added for some JSON functions at https://github.com/risingwavelabs/risingwave/issues/13016 but I think it was not generalized for any variadic function support?

### Describe the solution you'd like

PostgreSQL-compatible support of the `VARIADIC` keyword when calling functions with variadic parameters.

### Describe alternatives you've considered

I can't think of a cleaner-looking alternative than what PostgreSQL is doing.

For my immediate problem, this, um... finally does the trick.... 😞 

```
thumbtack=> select replace(
                replace(
                    replace(
                        replace(
                            regexp_replace(
                                replace(
                                    replace(
                                        'P1Y2M3DT4H5M6.15S', 'P', ''
                                    ), 'Y', ' year '
                                ), 'M', ' month '
                            ), 'DT', ' day '
                        ), 'H', ':'
                    ), 'M', ':'
                ), 'S', ''
            )::interval;
             replace
----------------------------------
 1 year 2 mons 3 days 04:05:06.15
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

resolve #14705

This PR adds VARIADIC argument support for the following functions:

```sql
select format('%s %s', variadic array['Hello', 'World']); -> "Hello World"
select concat(variadic array['abcde', '2', NULL, '22']); -> "abcde222"
select concat_ws(',', variadic array['abcde', 2, NULL, 22] :: varchar[]); -> "abcde,2,22"
select jsonb_build_array(variadic array[1, 2, 4, 5]); -> [1, 2, 4, 5]
select jsonb_build_object(variadic array['foo', '1', '2', 'bar']); -> {"2": "bar", "foo": 1}
select jsonb_extract_path('{"a": {"b": ["foo","bar"]}}', variadic array['a', 'b', '1']); -> bar
select jsonb_extract_path_text('{"a": {"b": ["foo","bar"]}}', variadic array['a', 'b', '1']); -> bar
```

It adds new functions `format_variadic`, `concat_ws_variadic`... on the backend, which take a fixed `anyarray` as the last argument. On the frontend, we rewrite the variadic calls into new functions.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

support VARIADIC arguments for the following functions:

- format
- concat_ws
- jsonb_build_array
- jsonb_build_object
- jsonb_extract_path
- jsonb_extract_path_text

The VARIADIC keyword expands the array parameter into one or more occurrences of its element type. See [postgres documentation](https://www.postgresql.org/docs/current/xfunc-sql.html#XFUNC-SQL-VARIADIC-FUNCTIONS) for more details.

Add a new function `concat`:

```
concat ( val1 "any" [, val2 "any" [, ...] ] ) → text
Concatenates the text representations of all the arguments. NULL arguments are ignored.
concat('abcde', 2, NULL, 22) → abcde222
```

### CODE REVIEW

File: proto/expr.proto [hunk 1]
```diff
@@ -224,9 +228,11 @@ message ExprNode {
     // jsonb ->> int, jsonb ->> text that returns text
     JSONB_ACCESS_STR = 601;
     // jsonb #> text[] -> jsonb
-    JSONB_EXTRACT_PATH = 613;
+    JSONB_EXTRACT_PATH = 626;
+    JSONB_EXTRACT_PATH_VARIADIC = 613;
```
Comment (xxchan) [inline proto/expr.proto:232]: Could you explain the safety of this change?

File: proto/expr.proto [hunk 2]
```diff
@@ -100,6 +101,8 @@ message ExprNode {
     CHAR_LENGTH = 225;
     REPEAT = 226;
     CONCAT_OP = 227;
+    CONCAT = 286;
+    CONCAT_VARIADIC = 287;
```
Comment (xxchan) [inline proto/expr.proto:105]: `concat` is a new function right? (It seems release note isn't complete

File: src/common/src/array/list_array.rs [hunk 3]
```diff
@@ -612,6 +612,24 @@ impl Debug for ListRef<'_> {
     }
 }
 
+impl Row for ListRef<'_> {
```
Comment (xxchan) [inline src/common/src/array/list_array.rs:615]: Is this a little confusing 🤔

File: proto/expr.proto (hunk 1 above)
Comment (xxchan) [reply proto/expr.proto:232]: So old impl is already variadic?
image

**Review by xxchan — COMMENTED** [review]
I'm still a little confused about why we need 2 exprs. Isn't variadic version converted to the variadic version? 🤔

Could you add some planner tests to help understand?

File: proto/expr.proto (hunk 2 above)
Comment (wangrunji0408) [reply proto/expr.proto:105]: Yes. Thanks for reminding.

File: src/common/src/array/list_array.rs (hunk 3 above)
Comment (wangrunji0408) [reply src/common/src/array/list_array.rs:615]: This is to allow arguments of list type to be accepted as `impl Row` in Rust functions.

File: src/expr/impl/src/scalar/concat.rs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+/// select concat('abcde', 2, NULL, 22);
+/// ----
+/// abcde222
+///
+/// query T
+/// select concat(variadic array['abcde', '2', NULL, '22']);
+/// ----
+/// abcde222
+/// ```
+#[function("concat(variadic anyarray) -> varchar")]
```
Comment (BugenZhao) [inline src/expr/impl/src/scalar/concat.rs:36]: I'm wondering what the possible receiver types are here. 👀

File: src/expr/impl/src/scalar/concat.rs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+fn concat(vals: impl Row, writer: &mut impl Write) {
```
Comment (BugenZhao) [inline src/expr/impl/src/scalar/concat.rs:37]: I'm wondering what the possible receiver types are here for `impl Row`. 👀

Comment (wangrunji0408) [reply src/expr/impl/src/scalar/concat.rs:37]: It is `RowRef` for non-variadic version, and `ListRef` for variadic version.

File: src/common/src/array/list_array.rs (hunk 3 above)
Comment (BugenZhao) [reply src/common/src/array/list_array.rs:615]: I'm okay with this as I cannot come up with another way to interpret `ARRAY` as rows to form an ambiguousness. If there's one, making a newtype on `ListRef` then `impl Row` on that will work.

**Review by BugenZhao — APPROVED** [review]
LGTM

**Pull request merged**